### ISSUE TITLE
Backend layer

### ISSUE THREAD

**Issue by hrach** [nextras/orm#72]
Currently, there is Nette Database, which is pretty good, but there are some missing parts. Like bulding more robust query (with manual joins), etc. I'm thinking about replacing Nette\Database.

What ORM needs:
- sql builder tool
- reflection of database (foreign keys)
- common layer for querying

Posibilities:
- **NDB fork**
  - **+** add what is needed, but had no meaning for NDB
  - **+** remove useless: type casting & conversion, "smart" column selection, smart relationships, ...
  - **-** no compatibility with NDB, just fork
  - **-** unpreffered license
- **create own solution**
  - **+** definitiely it will suits the best, no overhead, feateres as needed.
  - **-** but much work...
- **Doctrine\DBAL**
  - **+** very much awesome & tested work
  - **+** support for almost all drivers!
  - **+** it's known, pepople wouldn't be probably so much scared
  - **+** the right license
  - **-**  too much code
  - **-** "wired" solutions, like events
- **Stay with NDB** 
  - **+** advertising Nette
  - **+** reasonable size
  - **-** current SqlBuilder or Selection are far away what we need.
  - **-** totally useless features
- **dibi**
  - **+** small
  - **-** no tests
  - **-** quite unsufficient builder layer (missing things like addOrderBy, vs. setOrderBy)

For now, I pretty much like Doctrine\DBAL. What dou you think, guys? Dibi seems pretty usable too, but would require own SqlBuilder.

**Comment by fprochazka** [comment]
I'd be happy to extract DbalExtension to another repository, so we can share the code.

**Comment by JanTvrdik** [comment]
Why is dibi not an option?

**Comment by fprochazka** [comment]
> "wired" solutions, like events

You don't like the events implementation? I'm not sure libs like events can be done interchangably without somebody creating a standard for it. Which I'm strongly against. 

Kdyby/Events should be compatible with Nette events, Symfony Events and Doctrine Events :)

**Comment by fprochazka** [comment]
@JanTvrdik I think it's relevant to [Schema](https://github.com/doctrine/dbal/tree/master/lib/Doctrine/DBAL/Schema) and [Supported platforms](https://github.com/doctrine/dbal/tree/master/lib/Doctrine/DBAL/Platforms) which might be a bit bigger piece of code, but ORM should have features like this... and in Doctrine\DBAL, they're already implemented :)

**Comment by hrach** [comment]
How can I forgot about dibi? :( Added. And the truth is, it's probably relevant alternative.

**Comment by JanTvrdik** [comment]
And there are already two ORMs based on dibi (PetrP and LeanMapper) so it should be possible to based Nextras\Orm on it as well.

**Comment by fprochazka** [comment]
Btw, I don't think "too much code" is relevant. The lib might be big, but in reality, you're only picking what you want to use and ignore the rest.

**Comment by JanTvrdik** [comment]
Also I used Doctrine\DBAL in the past and it was pretty shitty experience. Mostly because it (from what I remember) uses prepared statements.

**Comment by hrach** [comment]
What are the (real) concerns about prepared statements (/emulated)?

**Comment by JanTvrdik** [comment]
@hrach It difficult to write queries with them. Doctrine DBAL has some workarounds but it still wasn't anywhere near as comfortable as dibi. Emulated prepared statements are useless as they are not really prepared statements and real prepared statements are slower in practice. See also http://www.zdrojak.cz/clanky/sql-injection-pre-kazdeho/?show=comments#comment-26337

**Comment by hrach** [comment]
I just realized that dibi has in fact no tests... so changeing well tested. Dibi can break almost every commit and nobody will know.

To statements - ndb use them too and it  wesn't be a problem for me at all.

**Comment by hrach** [comment]
Tody, I tried two things:
- write own solution: after about 30 minutes I discovered I'm rewriting Doctrine. (I had opened ndb, dibi and doctrine for inspiration).
- so I tried to rewrite support for Doctrine. I didn't finish, mostly because I stuck with binding args to placeholders - QueryBuilder in doctrine has args separated from calling where, order, etc. And these args has to be numbered when are setting to the query builder. Not an ideal solution.

So... I'm not so satisfied with Doctrine, however, I have to admint, that it isn't so bad. However, the documentation is quite bad :|
